Improbable/Stdlib/Improbable.Stdlib/WorkerConnection.cs

```diff
@@ -43,41 +44,59 @@ public string WorkerId
+            lock (connectionLock)
+            {
+                StopSendingMetrics();
+                CancelCommands();
+
+                connection?.Dispose();
+                connection = null;
+            }
 
-            connection?.Dispose();
-            connection = null;
         }
 
-        public static Task<WorkerConnection> ConnectAsync(IWorkerOptions workerOptions, ConnectionParameters connectionParameters, TaskCreationOptions taskOptions = TaskCreationOptions.None)
+        public static Task<WorkerConnection> ConnectAsync(IWorkerOptions workerOptions, ConnectionParameters connectionParameters, CancellationToken cancellation = default)
         {
             switch (workerOptions)
             {
                 case IReceptionistOptions receptionistOptions:
                     var workerName = workerOptions.WorkerName ?? $"{connectionParameters.WorkerType}-{Guid.NewGuid().ToString()}";
-                    return ConnectAsync(receptionistOptions.SpatialOsHost, receptionistOptions.SpatialOsPort, workerName, connectionParameters, taskOptions);
+                    return ConnectAsync(receptionistOptions.SpatialOsHost, receptionistOptions.SpatialOsPort, workerName, connectionParameters, cancellation);
 
                 case ILocatorOptions locatorOptions:
                     connectionParameters.Network.UseExternalIp = true;
-                    return ConnectAsync(locatorOptions, connectionParameters, taskOptions);
+                    return ConnectAsync(locatorOptions, connectionParameters, cancellation);
 
                 default:
                     throw new NotImplementedException("Unrecognized option type: " + workerOptions.GetType());
             }
         }
 
-        public static Task<WorkerConnection> ConnectAsync(string host, ushort port, string workerName, ConnectionParameters connectionParameters, TaskCreationOptions taskOptions = TaskCreationOptions.None)
+        public static Task<WorkerConnection> ConnectAsync(string host, ushort port, string workerName, ConnectionParameters connectionParameters, CancellationToken cancellation = default)
         {
-            var tcs = new TaskCompletionSource<WorkerConnection>(taskOptions);
+            var tcs = new TaskCompletionSource<WorkerConnection>();
+            if (cancellation.CanBeCanceled)
+            {
+                cancellation.Register(() => tcs.TrySetCanceled(cancellation));
+            }
 
             Task.Factory.StartNew(() =>
             {
                 try
                 {
                     using (var future = Connection.ConnectAsync(host, port, workerName, connectionParameters))
                     {
-                        var connection = future.Get();
+                        Connection connection;
+                        while (true)
+                        {
+                            cancellation.ThrowIfCancellationRequested();
+
+                            if (future.TryGet(out connection, 50))
```

Comment:
Ugh this "future" should _really_ be a `Task<Connection>` and this polling should be pushed back into the SDK imo